doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -438,25 +497,50 @@ make it impossible to implement this generic fallback behavior for
 ``__array_function__``.
 
 We could resolve this issue by change the handling of return values in
-``__array_function__`` in either of two possible ways: 1. Change the
-meaning of all arguments returning ``NotImplemented`` to indicate that
-all arguments should be coerced to NumPy arrays instead. However, many
-array libraries (e.g., scipy.sparse) really don't want implicit
-conversions to NumPy arrays, and often avoid implementing ``__array__``
-for exactly this reason. Implicit conversions can result in silent bugs
-and performance degradation. 2. Use another sentinel value of some sort
-to indicate that a class implementing part of the higher level array API
-is coercible as a fallback, e.g., a return value of
-``np.NotImplementedButCoercible`` from ``__array_function__``.
-
-If we take this second approach, we would need to define additional
-rules for how coercible array arguments are coerced, e.g., - Would we
-try for ``__array_function__`` overloads again after coercing coercible
-arguments? - If so, would we coerce coercible arguments one-at-a-time,
-or all-at-once?
-
-These are slightly tricky design questions, so for now we propose to
-defer this issue. We can always implement
+``__array_function__`` in either of two possible ways:
+
+1. Change the meaning of all arguments returning ``NotImplemented`` to indicate
+   that all arguments should be coerced to NumPy arrays instead. However, many
+   array libraries (e.g., scipy.sparse) really don't want implicit conversions
+   to NumPy arrays, and often avoid implementing ``__array__`` for exactly this
+   reason. Implicit conversions can result in silent bugs and performance
+   degradation.
```

Comment:
So what would happen if the policy were "if everyone returned `NotImplemented` then do what we used to do, and try coercing with `numpy.asarray`, if that fails, then we fail as we did before"?

Reply:
Yes, we could do this. I think this would a slightly inferior version of @mhvk's proposal to only try coercing classes that return `NotImplemented` if they implement `__array__`. The problem is that classes like `scipy.sparse.matrix` technically can be converted with `np.asarray()`, but you end up with a 0-dimensional array containing a sparse matrix (https://github.com/scipy/scipy/issues/4239). (That's probably a bug that could be fixed by defining an `__array__` method that raises TypeError.)